README.md
# daarem

NEWS
Log of changes to daarem

   o	first release (version 0.2):  August 6, 2018

   o	second release (version 0.3):  February, 2019
		- fixed a bug in how the default order was determined
		- fixed code for finding damping parameter; it can now handle the case 
		  of one or more singular values that are numerically zero 
		  
   o  third release (version 0.4):  September, 2019
	  - daarem can now handle cases where no objective function is 
	    provided by the user and only the fixed point function is provided
	    
   o  fourth release (version 0.4.1): December, 2019
    - no updates to the functionality of daarem; only modified a
      small chunk of related to checking a matrix class object
   
   o  fifth release (version 0.5): June, 2020
    - added option to determine convergence through differences in
      the value of the objective function. One sets convtype="objfn"
      if an objective function-based convergence criterion is deseired

R/fpiter.R
fpiter <- function(par, fixptfn, objfn=NULL, control=list( ), ...){

  control.default <- list(tol=1.e-07, maxiter=5000, trace=FALSE)
  namc <- names(control)
  if (!all(namc %in% names(control.default)))
    stop("unknown names in control: ", namc[!(namc %in% names(control.default))])
  ctrl <- modifyList(control.default, control)

  #
  # method = reduced-rank ("rre") or minimal-polynomial ("mpe") extrapolation
  # K = order of extrapolation scheme; K=2,3,4 are typical choices
  # square = a logical variable indicating whether or not "squaring" is used

  tol <- ctrl$tol
  maxiter <- ctrl$maxiter
  trace <- ctrl$trace

  if (trace) cat("fpiter \n")

  iter <- 1
  resid <- rep(NA,1)
  objeval <- 0
  conv <- FALSE

  while (iter < maxiter) {

    p.new <- fixptfn(par, ...)
    res <- sqrt(crossprod(p.new - par))

    if ( res < tol) {conv <- TRUE; break}

    if (trace) {
      if (!is.null(objfn)) {cat("Iter: ", iter, "Objective fn: ",objfn(par, ...), "\n"); objeval <- objeval + 1}
      else cat("Iter: ", iter, "Residual: ",res, "\n")
    }
    par <- p.new
    iter <- iter+1
  }

  loglik.best <-  if (!is.null(objfn)) objfn(par, ...) else NA

  return(list(par=par, value.objfn=loglik.best, fpevals=iter, objfevals = objeval, convergence=conv))
}


R/daarem_beta.R
daarem_beta <- function(par, fixptfn, objfn, ..., control=list()) {

  control.default <- list(maxiter=2000, order=10, tol=1.e-08, mon.tol=0.01, cycl.mon.tol=0.0, kappa=25, alpha=1.2)
  namc <- names(control)
  if (!all(namc %in% names(control.default))) {
    stop("unknown names in control: ", namc[!(namc %in% names(control.default))])
  }
  control <- modifyList(control.default, control)

  maxiter <- control$maxiter
  tol <- control$tol
  mon.tol <- control$mon.tol  ## monotonicity tolerance
  cycl.mon.tol <- control$cycl.mon.tol
  a1 <- control$alpha
  kappa <- control$kappa

  num.params <- length(par)
  nlag <- min(control$order, ceiling(num.params/2))

  Fdiff <- Xdiff <- matrix(0.0, nrow=num.params, ncol=nlag)
  obj_funvals <- rep(NA, maxiter + 2)

  xold <- par
  xnew <- fixptfn(xold, ...)
  obj_funvals[1] <- objfn(xold, ...)
  obj_funvals[2] <- objfn(xnew, ...)
  likchg <- obj_funvals[2] - obj_funvals[1]
  obj.evals <- 2

  fold <- xnew - xold
  k <- 1
  count <- 0
  shrink.count <- 0
  shrink.target <- 1/(1 + a1^kappa)
  lambda.ridge <- 100000
  r.penalty <- 0
  conv <- TRUE
  #num.em <- 0  ## number of EM fallbacks
  ell.star <- obj_funvals[2]
  while(k < maxiter) {
    count <- count + 1

    fnew <- fixptfn(xnew, ...) - xnew
    ss.resids <- sqrt(crossprod(fnew))
    if(ss.resids < tol & count==nlag) break

    Fdiff[,count] <- fnew - fold
    Xdiff[,count] <- xnew - xold

    np <- count
    if(np==1) {
        Ftmp <- matrix(Fdiff[,1], nrow=length(fnew), ncol=np)
        Xtmp <- matrix(Xdiff[,1], nrow=length(fnew), ncol=np)  ## is this matrix function needed?
    
        pow <- kappa - shrink.count
        s.delta <- exp(-0.5*log1p(a1^pow))
        dsq <- sum(Fdiff[,count]*Fdiff[,count])
        #dsq <- sum(Ftmp^2)
        lambda.ridge <- dsq*(1/s.delta - 1)
        ## What if dsq = 0?  
        gamma_vec <- sum(Fdiff[,count]*fnew)/(dsq + lambda.ridge)  
   
        r.penalty = 0
    } else {
        Ftmp <- Fdiff[,1:np]
        Xtmp <- Xdiff[,1:np]  

        tmp <- La.svd(Ftmp)
        dvec <- tmp$d
        dvec.sq <- dvec*dvec
        uy <- crossprod(tmp$u, fnew)
        uy.sq <- uy*uy

        ### Still need to compute Ftf
        Ftf <- sqrt(sum(uy.sq*dvec.sq))
        tmp_lam <- DampingFind(uy.sq, dvec, a1, kappa, shrink.count, Ftf, lambda.start=lambda.ridge, r.start=r.penalty)
        lambda.ridge <- tmp_lam$lambda
        r.penalty <- tmp_lam$rr
        dd <- (dvec*uy)/(dvec.sq + lambda.ridge)
        gamma_vec <- crossprod(tmp$vt, dd)
    }
    
    if(class(gamma_vec) != "try-error"){

      xbar <- xnew - drop(Xtmp%*%gamma_vec)
      fbar <- fnew - drop(Ftmp%*%gamma_vec)

      x.propose <- xbar + fbar
      new.objective.val <- try(objfn(x.propose, ...), silent=TRUE)
      obj.evals <- obj.evals + 1

      if(class(new.objective.val) != "try-error" & !is.na(obj_funvals[k+1]) &
         !is.nan(new.objective.val)) {
        if(new.objective.val >= obj_funvals[k+1] - mon.tol) {
          ## Increase delta
          obj_funvals[k+2] <- new.objective.val
          fold <- fnew
          xold <- xnew

          xnew <- x.propose
          shrink.count <- shrink.count + 1
        } else {
          ## Keep delta the same
          fold <- fnew
          xold <- xnew

          xnew <- fold + xold
          obj_funvals[k+2] <- objfn(xnew, ...)
          obj.evals <- obj.evals + 1
          #num.em <- num.em + 1
        }
      } else {
        ## Keep delta the same
        fold <- fnew
        xold <- xnew

        xnew <- fold + xold
        obj_funvals[k+2] <- objfn(xnew, ...)
        obj.evals <- obj.evals + 1
        count <- 0
        #num.em <- num.em + 1
      }
    } else {
      ## Keep delta the same
      fold <- fnew
      xold <- xnew

      xnew <- fold + xold
      obj_funvals[k+2] <- objfn(xnew, ...)
      obj.evals <- obj.evals + 1
      count <- 0
      #num.em <- num.em + 1
    }
    if(count==nlag) {
      count <- 0
      ## restart count
      ## make comparison here l.star vs. obj_funvals[k+2]
      if(obj_funvals[k+2] < ell.star - cycl.mon.tol) {
        ## Decrease delta
        shrink.count <- max(shrink.count - nlag, -2*kappa)
      }
      ell.star <- obj_funvals[k+2]
    }

    shrink.target <-  1/(1 + a1^(kappa - shrink.count))
    k <- k+1
  }
  obj_funvals <- obj_funvals[!is.na(obj_funvals)]
  value.obj <- objfn(xnew, ...)
  if(k >= maxiter) {
    conv <- FALSE
    warning("Algorithm did not converge")
  }
  return(list(par=c(xnew), fpevals = k, value.objfn=value.obj, objfevals=obj.evals, convergence=conv, objfn.track=obj_funvals))
}


R/daarem_base_objfn.R
daarem_base_objfn <- function(par, fixptfn, objfn, maxiter, tol, mon.tol, 
                              cycl.mon.tol, a1, kappa, num.params, nlag, 
                              check.par.resid, ...) {

 
  #maxiter <- control$maxiter
  #tol <- control$tol
  #mon.tol <- control$mon.tol  ## monotonicity tolerance
  #cycl.mon.tol <- control$cycl.mon.tol
  #a1 <- control$alpha
  #kappa <- control$kappa

  #num.params <- length(par)
  #nlag <- min(control$order, ceiling(num.params/2))

  Fdiff <- Xdiff <- matrix(0.0, nrow=num.params, ncol=nlag)
  obj_funvals <- rep(NA, maxiter + 2)

  xold <- par
  xnew <- fixptfn(xold, ...)
  obj_funvals[1] <- objfn(xold, ...)
  obj_funvals[2] <- objfn(xnew, ...)
  likchg <- obj_funvals[2] - obj_funvals[1]
  obj.evals <- 2

  fold <- xnew - xold
  k <- 1
  count <- 0
  shrink.count <- 0
  shrink.target <- 1/(1 + a1^kappa)
  lambda.ridge <- 100000
  r.penalty <- 0
  conv <- TRUE
  ell.star <- obj_funvals[2]
  while(k < maxiter) {
     count <- count + 1

     fnew <- fixptfn(xnew, ...) - xnew
     ss.resids <- sqrt(crossprod(fnew))
     if(ss.resids < tol & check.par.resid) break
     #new.objective.val >= obj_funvals[k+1] - mon.tol
     
     Fdiff[,count] <- fnew - fold
     Xdiff[,count] <- xnew - xold

     np <- count
     if(np==1) {
        Ftmp <- matrix(Fdiff[,1], nrow=num.params, ncol=np)
        Xtmp <- matrix(Xdiff[,1], nrow=num.params, ncol=np)  ## is this matrix function needed?
    
     } else {
        Ftmp <- Fdiff[,1:np]
        Xtmp <- Xdiff[,1:np]  
     }
     tmp <- La.svd(Ftmp)
     dvec <- tmp$d
     dvec.sq <- dvec*dvec
     uy <- crossprod(tmp$u, fnew)
     uy.sq <- uy*uy

     ### Still need to compute Ftf
     Ftf <- sqrt(sum(uy.sq*dvec.sq))
     tmp_lam <- DampingFind(uy.sq, dvec, a1, kappa, shrink.count, Ftf, lambda.start=lambda.ridge, r.start=r.penalty)
     lambda.ridge <- tmp_lam$lambda
     r.penalty <- tmp_lam$rr
     dd <- (dvec*uy)/(dvec.sq + lambda.ridge)
     gamma_vec <- crossprod(tmp$vt, dd)
   
     xbar <- xnew - drop(Xtmp%*%gamma_vec)
     fbar <- fnew - drop(Ftmp%*%gamma_vec)

     x.propose <- xbar + fbar
     new.objective.val <- try(objfn(x.propose, ...), silent=TRUE)
     obj.evals <- obj.evals + 1

     if(class(new.objective.val)[1] != "try-error" & !is.na(obj_funvals[k+1]) &
        !is.nan(new.objective.val)) {
         if(new.objective.val >= obj_funvals[k+1] - mon.tol) {  ## just change this line in daarem_base_noobjfn
            ## Increase delta
            obj_funvals[k+2] <- new.objective.val
            fold <- fnew
            xold <- xnew

            xnew <- x.propose
            shrink.count <- shrink.count + 1
         } else {
            ## Keep delta the same
            fold <- fnew
            xold <- xnew

            xnew <- fold + xold
            obj_funvals[k+2] <- objfn(xnew, ...)
            obj.evals <- obj.evals + 1
         }
     } else {
        ## Keep delta the same
         fold <- fnew
         xold <- xnew

         xnew <- fold + xold
         obj_funvals[k+2] <- objfn(xnew, ...)
         obj.evals <- obj.evals + 1
         count <- 0
     }
     if(count==nlag) {
         count <- 0
         ## restart count
         ## make comparison here l.star vs. obj_funvals[k+2]
         if(obj_funvals[k+2] < ell.star - cycl.mon.tol) {
            ## Decrease delta
            shrink.count <- max(shrink.count - nlag, -2*kappa)
         }
         ell.star <- obj_funvals[k+2]
     }
     if(abs(obj_funvals[k+2] - obj_funvals[k+1]) < tol & !check.par.resid) {
         break
     }
    shrink.target <-  1/(1 + a1^(kappa - shrink.count))
    k <- k+1
  }
  obj_funvals <- obj_funvals[!is.na(obj_funvals)]
  value.obj <- objfn(xnew, ...)
  if(k >= maxiter) {
    conv <- FALSE
  }
  return(list(par=c(xnew), fpevals = k, value.objfn=value.obj, objfevals=obj.evals, convergence=conv, objfn.track=obj_funvals))
}


R/daarem_base_noobjfn.R
daarem_base_noobjfn <- function(par, fixptfn, maxiter, tol, resid.tol, 
                                a1, kappa, num.params, nlag, ...) {
  
  Fdiff <- Xdiff <- matrix(0.0, nrow=num.params, ncol=nlag)
  rho <- resid.tol ## should this be user specified?
  
  resid_vals <- rep(NA, maxiter + 2)
  xold <- par
  xnew <- fixptfn(xold, ...)
  fold <- xnew - xold
  resid_vals[1] <- sqrt(crossprod(fold))
  fnew <- fixptfn(xnew, ...) - xnew
  resid_vals[2] <- sqrt(crossprod(fnew))
  ss.resids <- resid_vals[2]
  
  fp.evals <- 2
  k <- 1
  n.aa <- 0
  count <- 0
  shrink.count <- 0
  shrink.target <- 1/(1 + a1^kappa)
  lambda.ridge <- 100000
  r.penalty <- 0
  conv <- TRUE
  while(fp.evals < maxiter) {
     count <- count + 1
    
     Fdiff[,count] <- fnew - fold
     Xdiff[,count] <- xnew - xold
    
     np <- count
     if(np==1) {
        Ftmp <- matrix(Fdiff[,1], nrow=num.params, ncol=np)
        Xtmp <- matrix(Xdiff[,1], nrow=num.params, ncol=np)  
     } else {
        Ftmp <- Fdiff[,1:np]
        Xtmp <- Xdiff[,1:np]  
     }
     tmp <- La.svd(Ftmp)
     dvec <- tmp$d
     dvec.sq <- dvec*dvec
     uy <- crossprod(tmp$u, fnew)
     uy.sq <- uy*uy
    
     ### Still need to compute Ftf
     Ftf <- sqrt(sum(uy.sq*dvec.sq))
     tmp_lam <- DampingFind(uy.sq, dvec, a1, kappa, shrink.count, Ftf, lambda.start=lambda.ridge, r.start=r.penalty)
     lambda.ridge <- tmp_lam$lambda
     r.penalty <- tmp_lam$rr
     dd <- (dvec*uy)/(dvec.sq + lambda.ridge)
     gamma_vec <- crossprod(tmp$vt, dd)
    
     xbar <- xnew - drop(Xtmp%*%gamma_vec)
     fbar <- fnew - drop(Ftmp%*%gamma_vec)
     x.propose <- xbar + fbar
       
     f.propose <- fixptfn(x.propose, ...) - x.propose
     fp.evals <- fp.evals + 1
     ss.propose <- sqrt(crossprod(f.propose))
     if(ss.propose <= ss.resids*(1.00 + rho^k)) {  
         ## Increase delta
         fold <- fnew
         xold <- xnew
         xnew <- x.propose
         fnew <- f.propose
          
         shrink.count <- shrink.count + 1
         ss.resids <- ss.propose
         n.aa <- n.aa + 1
     } else {
         ## Keep delta the same
         fold <- fnew
         xold <- xnew
         xnew <- fold + xold
          
         fnew <- fixptfn(xnew, ...) - xnew
         ss.resids <- sqrt(crossprod(fnew))
         fp.evals <- fp.evals + 1
     }

     if(ss.resids < tol & count==nlag) break
    
     resid_vals[k + 2] <- ss.resids
     if(count==nlag) {
        count <- 0
        ## restart count
     }
     shrink.target <-  1/(1 + a1^(kappa - shrink.count))
     k <- k+1
  }
  if(fp.evals >= maxiter) {
    conv <- FALSE
  }
  return(list(par=c(xnew), fpevals = fp.evals, value.objfn=NULL, objfevals=NULL, convergence=conv, objfn.track=NULL,
              residuals=resid_vals[!is.na(resid_vals)], n.aa=n.aa))
}


R/daarem.R
daarem <- function(par, fixptfn, objfn, ..., control=list()) {

  control.default <- list(maxiter=2000, order=5, tol=1.e-08, mon.tol=0.01, cycl.mon.tol=0.0, kappa=25, 
                          alpha=1.2, resid.tol=0.95, convtype="param")
  namc <- names(control)
  if (!all(namc %in% names(control.default))) {
    stop("unknown names in control: ", namc[!(namc %in% names(control.default))])
  }
  control <- modifyList(control.default, control)

  maxiter <- control$maxiter
  tol <- control$tol
  mon.tol <- control$mon.tol  ## monotonicity tolerance
  cycl.mon.tol <- control$cycl.mon.tol
  a1 <- control$alpha
  kappa <- control$kappa
  resid.tol <- control$resid.tol
  if(control$convtype=="param") {
      check.par.resid <- TRUE
  } else if(control$convtype=="objfn") {
      check.par.resid <- FALSE
  }

  num.params <- length(par)
  nlag <- min(control$order, ceiling(num.params/2))

  if(!missing(objfn)) {
      ans <- daarem_base_objfn(par, fixptfn, objfn, maxiter, tol, mon.tol, 
                               cycl.mon.tol, a1, kappa, num.params, nlag, 
                               check.par.resid, ...) 
  } else {
      ans <- daarem_base_noobjfn(par, fixptfn, maxiter, tol, resid.tol, 
                                 a1, kappa, num.params, nlag, ...) 
  }
  if(!ans$convergence) {
     warning("Algorithm did not converge")
  }
  return(ans)
}
